Source/NonlinearSolvers/JacobianFunctionMF.H

```diff
@@ -0,0 +1,235 @@
+    inline
+    void create ( T& a_Z, const T& a_U )
+    {
+        a_Z.define(a_U);
+    }
+
+    T makeVecLHS () const;
+    T makeVecRHS () const;
+
+    inline
+    void assign( T& a_Z, const T& a_U ) {
+        a_Z = a_U;
+    }
+
+    inline
+    void increment( T& a_Z, const T& a_U, RT a_scale )
+    {
+        a_Z += a_scale*a_U;
+    }
+
+    inline
+    void scale ( T& a_U, RT a_scale )
+    {
+        a_U.scale(a_scale);
+    }
+
+    inline
+    void linComb ( T& a_U, RT a, const T& X, RT b, const T& Y )
+    {
+        a_U.linComb( a, X, b, Y );
+    }
+
+    inline
+    void setToZero ( T& a_U )
+    {
+        a_U.zero();
+    }
+
+    inline
+    void setVal ( T& a_U, RT a_val )
+    {
+        a_U.setVal(a_val);
+    }
+
+    inline
+    RT dotProduct( const T& a_X, const T& a_Y )
+    {
+        return( a_X.dotProduct(a_Y) );
+    }
+
+    inline
+    RT norm2( const T& a_U )
+    {
+        return ( a_U.norm2() );
+    }
+
+    [[nodiscard]] inline
+    bool isDefined() const { return m_is_defined;  }
+
+    inline
+    void setBaseSolution ( const T&  a_U )
+    {
+        m_Y0.Copy(a_U);
+        m_normY0 = norm2(m_Y0);
+    }
+
+    inline
+    void setBaseRHS ( const T&  a_R )
+    {
+        m_R0.Copy(a_R);
+    }
+
+    inline
+    void setJFNKEps ( RT a_eps )
+    {
+        m_epsJFNK = a_eps;
+    }
+
+    inline
+    void setIsLinear ( bool a_isLinear )
+    {
+        m_is_linear = a_isLinear;
+    }
+
+    inline
+    void curTime ( RT a_time )
+    {
+        m_cur_time = a_time;
+    }
+
+    inline
+    void curTimeStep ( RT a_dt )
+    {
+        m_dt = a_dt;
+    }
+
+    void define( const T&, Ops* );
+
+    private:
+
+    bool m_is_defined = false;
+    bool m_is_linear = false;
+    bool m_usePreCond = false;
+    RT m_epsJFNK = RT(1.0e-6);
+    RT m_normY0;
+    RT m_cur_time, m_dt;
+    std::string m_pc_type;
+
+    T m_Z, m_Y0, m_R0, m_R;
+    Ops* m_ops;
+
+};
+
+template <class T, class Ops>
+void JacobianFunctionMF<T,Ops>::define ( const T&  a_U,
+                                         Ops*      a_ops )
+{
+    m_Z.Define(a_U);
+    m_Y0.Define(a_U);
+    m_R0.Define(a_U);
+    m_R.Define(a_U);
+
+    m_ops = a_ops;
+
+    m_is_defined = true;
+}
+
+template <class T, class Ops>
+auto JacobianFunctionMF<T,Ops>::makeVecRHS () const -> T
+{
+    T Vec;
+    Vec.Define(m_R);
+    return Vec;
+}
+
+template <class T, class Ops>
+auto JacobianFunctionMF<T,Ops>::makeVecLHS () const -> T
+{
+    T Vec;
+    Vec.Define(m_R);
+    return Vec;
+}
+
+template <class T, class Ops>
+void JacobianFunctionMF<T,Ops>::apply (T& a_dF, const T& a_dU)
+{
+    BL_PROFILE("JacobianFunctionMF::apply()");
+    using namespace amrex::literals;
+    RT const normY = norm2(a_dU); // always 1 when called from GMRES
+
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+        isDefined(),
+        "JacobianFunction::apply() called on undefined JacobianFunction");
+
+    if (normY < 1.0e-15) { a_dF.zero(); }
+    else {
+
+        RT eps;
```

Comment:
The calculation of `eps` should be moved outside the apply function and it can be done before GMRES begins. Although it changes for each Newton iteration, it should be the same for GRMRES iterations. If it does change, the system for GMRES is no longer linear.

Reply:
It seems the matrix changes during the call of `GMRES::solve` because of the way how `apply` is implemented. Here are some observations. I added this to AMReX_GMRES.H
```
diff --git a/Src/LinearSolvers/AMReX_GMRES.H b/Src/LinearSolvers/AMReX_GMRES.H
index 5ccc68cd23..0404316a67 100644
--- a/Src/LinearSolvers/AMReX_GMRES.H
+++ b/Src/LinearSolvers/AMReX_GMRES.H
@@ -232,6 +232,10 @@ void GMRES<V,M>::solve (V& a_sol, V const& a_rhs, RT a_tol_rel, RT a_tol_abs, in
 
     if (m_status == -1 && m_its >= a_its) { m_status = 1; }
 
+    amrex::Print() << "   xxxxx    m_res = " << m_res << '\n';
+    compute_residual(m_vv[0], a_sol, a_rhs);
+    amrex::Print() << "   xxxxx ||b-Ax|| = " << m_linop->norm2(m_vv[0]) << '\n';
+
     m_v_tmp_rhs.reset();
     m_v_tmp_lhs.reset();
     m_vv.clear();
```
This prints out `m_res`, which is what GMRES thinks the final residual is, and the residual computed directly from the final solution and RHS. They should agree. But instead I get
```
Newton: iteration =   0, norm = 8.12653e+12 (abs.), 1.00000e+00 (rel.)
GMRES: iter = 0, residual = 8.126534683e+12, 1 (rel.)
GMRES: iter = 1, residual = 6.101812917e+12, 0.7508505354 (rel.)
GMRES: iter = 2, residual = 5.206856998e+11, 0.06407229159 (rel.)
GMRES: iter = 3, residual = 5.739630096e+10, 0.007062826064 (rel.)
GMRES: iter = 4, residual = 6795482960, 0.0008362091871 (rel.)
GMRES: iter = 5, residual = 1080455821, 0.0001329540651 (rel.)
GMRES: iter = 6, residual = 165349353.5, 2.034684647e-05 (rel.)
   xxxxx    m_res = 165349353.5
   xxxxx ||b-Ax|| = 4.870637847e+12
```
The only explanation seems to be A (i.e., the apply function) has changed during the life of one GMRES::solve call. Note that I am not saying A should not change from one Newton iteration to another (i.e., from one GMRES::solve call to another), but within one Newton iteration, the matrix needs to be same (i.e., apply needs to represent the same matrix).

For comparison, I run amrex/Tests/LinearSolvers/CurlCurl and get
```
GMRES: iter = 0, residual = 33321.69994, 1 (rel.)
GMRES: iter = 1, residual = 4207.924549, 0.126281809 (rel.)
GMRES: iter = 2, residual = 48.69181953, 0.00146126457 (rel.)
GMRES: iter = 3, residual = 0.2033884438, 6.103783544e-06 (rel.)
GMRES: iter = 4, residual = 0.002315450854, 6.948777697e-08 (rel.)
GMRES: iter = 5, residual = 6.717085573e-05, 2.0158292e-09 (rel.)
GMRES: iter = 6, residual = 2.859065516e-06, 8.58019105e-11 (rel.)
   xxxxx    m_res = 2.859065516e-06
   xxxxx ||b-Ax|| = 2.859064266e-06
```

Reply:
I'm not sure what input deck you are using. If you are using Marco's, which we know is not converging and has issues, then that may explain this large difference. Two more comments below

1) Since the norm of the vector being passed by AMRex::GMRES to WarpX::JacobianFunctionMF::apply() is always pre-normalized to unity, then the epsilon used by that routine does not change at all during the linear iterations.
2) There is no need for epsilon to remain strictly fixed during GMRES. A finite-difference approximation to J*v will always have some error when the function is nonlinear. The error will scale with epsilon * norm(v). Epsilon needs to be chosen such that the error is small.

Reply:
I am using Marco's input deck. My point is the issue might be the matrix is changing during a single GMRES. It's like GMRES is give a problem Ax = b. During the middle of its calculation, it switches to solve Mx = b, where M is different from M and GMRES does not even know about because the matrix is not owned by it. In the end, we get something is a solution of neither Ax = b nor Mx = b.

1. It's possible eps is not the issue. Then there is something else in apply that changes the matrix. We never explicitly construct the matrix A, but conceptually we still have a matrix and we use it in the apply function. 
2. GMRES is based subspace span{r, Ar, A^2 r, A^3 r, ...}. If A changes, the space becomes span{r, Ar, A A_1 r, A A_1 A_2 r, ...}, where A, A_1, A_2, ... can be different. I am not sure what to expect.

Reply:
There is also the performance consideration. The norm calculation involves mpi allreduce. If it produces the same result everytime apply is called, it would be better to avoid calling mpi allreduce repeatedly.

Reply:
Are there Dirichlet boundaries in Marco's test? If so, we need to zero out the RHS on the Dirichlet edges. This is because the norm calculation does not exclude those edges.

Reply:
I think you are correct in that A is changing when it is not meant to. It seemed to me during the meeting on Friday that the issue has to do with special particles for lasers that are unintentionally being updated each iteration in the JFNK process.